### PR TITLE
fix issue #123

### PR DESCRIPTION
**Pull request by jinfeihan57**
using real_filename() to fix issue #123

### CODE REVIEW

File: src/zip.c [hunk 1]
```diff
@@ -875,6 +921,11 @@ static int extract(mz_zip_archive *zip_archive, const char *dir,
       // Cannot get information about zip archive;
       goto out;
     }
+    if (!real_filename(info.m_filename, info.m_filename,
```
Comment (jinfeihan57) [inline src/zip.c:924]: info.m_filename is a array not a ptr. so API real_filename can't return ptr.
name realpath() is already be taken, so I use real_filename() or real_path(). WDT?

**Comment by jinfeihan57** [comment]
@kuba-- 
Any suggestions on mac build ERROR?

**Comment by kuba--** [comment]
> @kuba--
> Any suggestions on mac build ERROR?

I think it's related to:
```
analyze-cc: INFO: /Users/runner/work/zip/zip/src/zip.c:853:9: warning: Value stored to 'src' is never read
16
analyze-cc: INFO:         src += 1;
17
analyze-cc: INFO:         ^      ~
18
analyze-cc: INFO: 1 warning generated.
19
```

**Comment by kuba--** [comment]
Look at the cases, like:
```
"../a/b/c/../"  

"../"
```

**Comment by kuba--** [comment]
Generally I think we can simplify it a little bit.
I quickly did some _exercise_ and wrote something like:
```c

char *simplify(const char *str, size_t len) {
  char *out = (char *)malloc(len * sizeof(char));
  memset(out, 0, len);

  size_t offstr = 0;
  size_t offout = 0, ncopied = 0;

  // skip trailing '/'
  while(ISSLASH(str[offstr])) offstr++;

  for (; offstr < len; offstr++) {
    if (ISSLASH(str[offstr])) {
      if (ncopied > 0 && strncmp(&out[offout], ".", 1) && strncmp(&out[offout], "..", 2)) {
        offout += ncopied;
        out[offout++] = str[offstr]; // append '/'
      }
      ncopied = 0;
    } else {
      out[offout + ncopied] = str[offstr];
      ncopied++;
    }
  }

  // at the end, extra check what we've already copied
  if (ncopied == 0 || !strncmp(&out[offout], ".", 1) || !strncmp(&out[offout], "..", 2)) {
    out[offout] = 0;
  }

  return out;
}
```

I think it works fine, you can take a look.

**Comment by jinfeihan57** [comment]
I fix my BUG.
and I test your func. they all working fine.
but I think your function have to malloc memory space，when it's all done, you have to free it.
I think it's too cumbersome to use.

**Comment by kuba--** [comment]
> I fix my BUG.
> and I test your func. they all working fine.
> but I think your function have to malloc memory space，when it's all done, you have to free it.
> I think it's too cumbersome to use.

Actually it's not a problem. You can just easily remove first two lines (malloc and memset) and return to the passed string (I just wanted to quickly test it out and compare), so we may have:
```c
char *simplify(char *str, char *out, size_t len) {
  size_t offstr = 0;
  size_t offout = 0, ncopied = 0;

  // skip trailing '/'
  while(ISSLASH(*str)) str++;

	for (; offstr < len; offstr++) {
    if (ISSLASH(str[offstr])) {
      if (ncopied > 0 && strncmp(&out[offout], ".", 1) && strncmp(&out[offout], "..", 2)) {
        offout += ncopied;
        out[offout++] = str[offstr]; // append '/'
      }
      ncopied = 0;
    } else {
      out[offout + ncopied] = str[offstr];
      ncopied++;
    }
  }

  // at the end, extra check what we've already copied
  if (ncopied == 0 || !strncmp(&out[offout], ".", 1) || !strncmp(&out[offout], "..", 2)) {
    out[offout] = 0;
  }
  return out;
}
```

Also take a look into following case, where you have a '.' at the end of filename (like missed extension), e.g.: ` "/a./b."`, or `"/a./b..../"`
It does not happen very often but it's a correct path (apart of from trailing '/').
So, I would expect to get `"a./b."`, `"a./b..../"`.

**Comment by jinfeihan57** [comment]
Ok,
But "a./b." or "a./b..../" are illegal in linux system.

**Comment by jinfeihan57** [comment]
@kuba-- 
I only added arg verify.

**Comment by kuba--** [comment]
> Ok,
> But "a./b." or "a./b..../" are illegal in linux system.

what do you mean _illegal_ you may have files like that - no problems
```
echo "whatever" > "a." 
touch "b....."
```

**Comment by kuba--** [comment]
> @kuba--
> I only added arg verify.

OK, thanks.
So if we wanna use it then we need to rename couple things (I think):
```
static char *normalize(char *name, char *const nname, size_t len) 
```

Also offsets can be renamed just to `offn, offnn, ncpy`.

And we also need some test for that.

**Review by kuba-- — APPROVED** [review]

**Pull request merged**